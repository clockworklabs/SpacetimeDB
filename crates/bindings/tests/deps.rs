//! Snapshot testing for the dependency tree of the `bindings` crate - we want
//! to make sure we don't unknowingly add a bunch of dependencies here,
//! slowing down compilation for every spacetime module.

// We need to remove the `cpufeatures` and `libc` dependencies from the output, it added on `arm` architecture:
// https://github.com/RustCrypto/sponges/blob/master/keccak/Cargo.toml#L24-L25, breaking local testing.
#[cfg(target_arch = "aarch64")]
fn hack_keccack(cmd: String) -> String {
    let mut found = false;
    let mut lines = cmd.lines().peekable();
    let mut output = String::new();

    while let Some(line) = lines.next() {
        // Check we only match keccak/cpufeatures/libc
        if line.contains("keccak") {
            found = true;
        }
        if found && line.contains("cpufeatures") {
            if let Some(next_line) = lines.peek() {
                if next_line.contains("libc") {
                    found = false;
                    // Skip libc
                    lines.next();
                    continue;
                }
            }
        }
        output.push_str(line);
        output.push('\n');
    }

    output
}
#[cfg(not(target_arch = "aarch64"))]
fn hack_keccack(cmd: String) -> String {
    cmd
}

#[test]
fn deptree_snapshot() -> std::io::Result<()> {
    let cmd = "cargo tree -p spacetimedb -f {lib} -e no-dev";
    let deps_tree = hack_keccack(run_cmd(cmd));
    let all_deps = hack_keccack(run_cmd("cargo tree -p spacetimedb -e no-dev --prefix none --no-dedupe"));
    let mut all_deps = all_deps.lines().collect::<Vec<_>>();
    all_deps.sort();
    all_deps.dedup();
    let num_deps = all_deps.len();

    insta::assert_snapshot!(
        "spacetimedb_bindings_dependencies",
        format!("total crates: {num_deps}\n{deps_tree}"),
        cmd
    );

    let cmd = "cargo tree -p spacetimedb -e no-dev -d --depth 0";
    insta::assert_snapshot!("duplicate_deps", run_cmd(cmd), cmd);

    Ok(())
}

// runs a command string, split on spaces
#[track_caller]
fn run_cmd(cmd: &str) -> String {
    let mut args = cmd.split(' ');
    let output = std::process::Command::new(args.next().unwrap())
        .args(args)
        .stdout(std::process::Stdio::piped())
        .output()
        .unwrap();
    assert!(output.status.success());
    String::from_utf8(output.stdout).unwrap()
}
