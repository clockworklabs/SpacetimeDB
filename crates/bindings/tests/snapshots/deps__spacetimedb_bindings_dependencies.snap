---
source: crates/bindings/tests/deps.rs
expression: "cargo tree -p spacetimedb -f {lib} -e no-dev"
---
total crates: 68
spacetimedb
├── bytemuck
├── derive_more
│   ├── convert_case
│   ├── proc_macro2
│   │   └── unicode_ident
│   ├── quote
│   │   └── proc_macro2 (*)
│   └── syn
│       ├── proc_macro2 (*)
│       ├── quote (*)
│       └── unicode_ident
│   [build-dependencies]
│   └── rustc_version
│       └── semver
├── getrandom
│   ├── cfg_if
│   └── libc
├── log
├── rand
│   ├── libc
│   ├── rand_chacha
│   │   ├── ppv_lite86
│   │   │   └── zerocopy
│   │   │       ├── byteorder
│   │   │       └── zerocopy_derive
│   │   │           ├── proc_macro2 (*)
│   │   │           ├── quote (*)
│   │   │           └── syn (*)
│   │   └── rand_core
│   │       └── getrandom (*)
│   └── rand_core (*)
├── scoped_tls
├── spacetimedb_bindings_macro
│   ├── heck
│   ├── humantime
│   ├── proc_macro2 (*)
│   ├── quote (*)
│   ├── spacetimedb_primitives
│   │   ├── bitflags
│   │   ├── either
│   │   ├── itertools
│   │   │   └── either
│   │   └── nohash_hasher
│   └── syn (*)
├── spacetimedb_bindings_sys
│   └── spacetimedb_primitives (*)
├── spacetimedb_lib
│   ├── anyhow
│   ├── bitflags
│   ├── blake3
│   │   ├── arrayref
│   │   ├── arrayvec
│   │   ├── cfg_if
│   │   └── constant_time_eq
│   │   [build-dependencies]
│   │   └── cc
│   │       └── shlex
│   ├── chrono
│   │   └── num_traits
│   │       [build-dependencies]
│   │       └── autocfg
│   ├── derive_more (*)
│   ├── enum_as_inner
│   │   ├── heck
│   │   ├── proc_macro2 (*)
│   │   ├── quote (*)
│   │   └── syn (*)
│   ├── hex
│   ├── itertools (*)
│   ├── rand (*)
│   ├── spacetimedb_bindings_macro (*)
│   ├── spacetimedb_data_structures
│   │   ├── ahash
│   │   │   ├── cfg_if
│   │   │   ├── getrandom (*)
│   │   │   ├── once_cell
│   │   │   └── zerocopy (*)
│   │   │   [build-dependencies]
│   │   │   └── version_check
│   │   ├── hashbrown
│   │   │   └── equivalent
│   │   ├── nohash_hasher
│   │   ├── smallvec
│   │   └── thiserror
│   │       └── thiserror_impl
│   │           ├── proc_macro2 (*)
│   │           ├── quote (*)
│   │           └── syn (*)
│   ├── spacetimedb_primitives (*)
│   ├── spacetimedb_sats
│   │   ├── anyhow
│   │   ├── arrayvec
│   │   ├── bitflags
│   │   ├── bytemuck
│   │   ├── bytes
│   │   ├── chrono (*)
│   │   ├── decorum
│   │   │   ├── approx
│   │   │   │   └── num_traits (*)
│   │   │   └── num_traits (*)
│   │   ├── derive_more (*)
│   │   ├── enum_as_inner (*)
│   │   ├── ethnum
│   │   │   └── serde
│   │   ├── hex
│   │   ├── itertools (*)
│   │   ├── second_stack
│   │   ├── sha3
│   │   │   ├── digest
│   │   │   │   ├── block_buffer
│   │   │   │   │   └── generic_array
│   │   │   │   │       └── typenum
│   │   │   │   │       [build-dependencies]
│   │   │   │   │       └── version_check
│   │   │   │   └── crypto_common
│   │   │   │       ├── generic_array (*)
│   │   │   │       └── typenum
│   │   │   └── keccak
│   │   ├── smallvec
│   │   ├── spacetimedb_bindings_macro (*)
│   │   ├── spacetimedb_primitives (*)
│   │   └── thiserror (*)
│   └── thiserror (*)
└── spacetimedb_primitives (*)
